lib/src/puppetdb-cli.cc

```diff
@@ -155,19 +159,45 @@ size_t write_body(char *ptr, size_t size, size_t nmemb, void *userdata){
     return size * nmemb;
 }
 
+string
+convert_query_to_post_data(const string& query_str) {
+    // If this is PQL then we need to wrap the query in double-quotes, otherwise
+    // the query is AST and we leave it alone
+    rapidjson::StringBuffer s;
+    rapidjson::Writer<rapidjson::StringBuffer> writer(s);
+    writer.StartObject();
+    writer.String("query");
+
+    string query_str_copy { query_str };
+    boost::trim(query_str_copy);
+    if (boost::starts_with(query_str_copy, "[")) {
+        rapidjson::StringStream is(query_str.c_str());
```

Comment:
[3/4] Why memory is hard in C

Here we're taking the C string (a char pointer to the char array that backs the C++ style string) which is what  [StringStream}(http://rapidjson.org/md_doc_stream.html#StringStream) requires.  At this point I wondered why not use the trimmed string (query_str_copy from above) or maybe a passed by value version of query_str. Seemed like either would cleanup the code some and not sending the leading spaces to PDB sounds like it's a decent idea.

Talking with @ajroetker he needed to make several attempts at this before getting the right semantics from a rapidjson perspective. The issue is what rapidjson is doing with query_str. Line 176 and line 179 are parsing query_str and "emitting" it to the writer that is backed by the rapidjson StringBuffer. We return the contents of that buffer at line 183 below. The value returned here is used where the [4/4] comment is found. If we use the copied string or a pass by value string into this method, the result is a bunch of null characters.

What rapidjson must be doing when we parse query_str is just copying over the char* from the StringStream on the input side to the StringBuffer on the output side. When we leave the method and it's no longer in scope, the char array that the pointer refers to is deallocated and results in null characters to be returned. This isn't exactly documented behavior, but might be related to "[in situ parsing]"(http://rapidjson.org/md_doc_dom.html#InSituParsing).